xds/internal/xdsclient/xdsresource/unmarshal_rds.go

```diff
@@ -313,29 +314,23 @@ func routesProtoToSlice(routes []*v3routepb.Route, csps map[string]clusterspecif
 				route.WeightedClusters[a.Cluster] = WeightedCluster{Weight: 1}
 			case *v3routepb.RouteAction_WeightedClusters:
 				wcs := a.WeightedClusters
-				var totalWeight uint32
+				var totalWeight uint64
 				for _, c := range wcs.Clusters {
 					w := c.GetWeight().GetValue()
 					if w == 0 {
 						continue
 					}
+					if (totalWeight + uint64(w)) > math.MaxUint32 {
+						return nil, nil, fmt.Errorf("total weight of clusters exceed MaxUint32")
+					}
 					wc := WeightedCluster{Weight: w}
 					cfgs, err := processHTTPFilterOverrides(c.GetTypedPerFilterConfig())
 					if err != nil {
 						return nil, nil, fmt.Errorf("route %+v, action %+v: %v", r, a, err)
 					}
 					wc.HTTPFilterConfigOverride = cfgs
 					route.WeightedClusters[c.GetName()] = wc
-					totalWeight += w
-				}
-				// envoy xds doc
-				// default TotalWeight https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#envoy-v3-api-field-config-route-v3-weightedcluster-total-weight
-				wantTotalWeight := uint32(100)
-				if tw := wcs.GetTotalWeight(); tw != nil {
-					wantTotalWeight = tw.GetValue()
-				}
-				if totalWeight != wantTotalWeight {
-					return nil, nil, fmt.Errorf("route %+v, action %+v, weights of clusters do not add up to total total weight, got: %v, expected total weight from response: %v", r, a, totalWeight, wantTotalWeight)
+					totalWeight += uint64(w)
```

Comment:
Minor nit. If you move this line to 322, then it will look simpler I feel:

```golang
totalWeight += uint64(w)
if totalWeight > math.MaxUint32 {
	return nil, nil, fmt.Errorf("xds: total weight of clusters exceeds MaxUint32")
}
```

Reply:
Agreed. updating..